pytorch_lightning/utilities/params_tying.py

```diff
@@ -0,0 +1,70 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Utilities for automatic parameters tying
+Reference:
+    https://github.com/pytorch/fairseq/blob/1f7ef9ed1e1061f8c7f88f8b94c7186834398690/fairseq/trainer.py#L110-L118
+"""
+from typing import Dict, List, Optional
+
+from torch.nn import Parameter
+
+import pytorch_lightning as pl
+
+
+def find_shared_parameters(
+    module: "pl.LightningModule", tied_parameters: Optional[Dict] = None, prefix: str = ""
+) -> List[str]:
```

Comment:
what's expected to be the public interface for callers? it seems like this is both the public API and the recursive helper. to avoid possible bugs by end users, we could split this into

1. `def find_shared_parameters(module: nn.Module) -> List[str]:`
2. `def _find_shared_parameters(module: nn.Module, tied_parameters: Optional[Dict] = None, prefix: str = "") -> List[str]: `

and the public API alls the helper, but doesn't expose the tied parameters or prefix. 

what do you think?

Reply:
Makes sense! better design. Thanks